src/ruby/ext/grpc/rb_channel.c

```diff
@@ -335,10 +367,118 @@ static VALUE grpc_rb_channel_get_target(VALUE self) {
   return res;
 }
 
+// Either start polling channel connection state or signal that it's free to
+// destroy.
+// Not safe to call while a channel's connection state is polled.
+static void grpc_rb_channel_try_register_connection_polling(
+    grpc_rb_channel *wrapper) {
+  grpc_connectivity_state conn_state;
+  gpr_timespec sleep_time = gpr_time_add(
+      gpr_now(GPR_CLOCK_REALTIME), gpr_time_from_millis(20, GPR_TIMESPAN));
+
+  GPR_ASSERT(wrapper);
+  GPR_ASSERT(wrapper->wrapped);
+  gpr_mu_lock(&wrapper->safe_destroy_mu);
+  if (wrapper->request_safe_destroy) {
+    wrapper->safe_to_destroy = 1;
+    gpr_cv_signal(&wrapper->safe_destroy_cv);
+    gpr_mu_unlock(&wrapper->safe_destroy_mu);
+    return;
+  }
+  gpr_mu_lock(&channel_polling_mu);
+  conn_state = grpc_channel_check_connectivity_state(wrapper->wrapped, 0);
+  // avoid posting work to the channel polling cq if it's been shutdown
+  if (!abort_channel_polling && conn_state != GRPC_CHANNEL_SHUTDOWN) {
+    grpc_channel_watch_connectivity_state(
+        wrapper->wrapped, conn_state, sleep_time, channel_polling_cq, wrapper);
+  } else {
+    wrapper->safe_to_destroy = 1;
+    gpr_cv_signal(&wrapper->safe_destroy_cv);
+  }
+  gpr_mu_unlock(&channel_polling_mu);
+  gpr_mu_unlock(&wrapper->safe_destroy_mu);
+}
+
+// Note requires wrapper->wrapped, wrapper->safe_destroy_mu/cv initialized
+static void grpc_rb_channel_safe_destroy(grpc_rb_channel *wrapper) {
+  gpr_mu_lock(&wrapper->safe_destroy_mu);
+  if (!wrapper->safe_to_destroy) {
+    wrapper->request_safe_destroy = 1;
+    gpr_cv_wait(&wrapper->safe_destroy_cv, &wrapper->safe_destroy_mu,
```

Comment:
This should be enclosed within a `while (!wrapper->safe_to_destroy)` loop to avoid problems with spurious wakeups.